### ISSUE TITLE
Release 2.0.0 Checklist

### ISSUE THREAD

**Issue by pboling** [oauth-xx/oauth2#332]
We have a new team of maintainers!

Myself (@pboling), @xtagon, @josephpage, @galois17 and a few others, are volunteering their efforts.

We are working on getting a new release out, @ *2.0.0*.  This is a checklist to make sure we do it in an orderly fashion.

- [x] What is the current minimum Ruby Version requirement of the gem, based on internal code and dependencies?
  - [x] [FIX] Documentation in `README`
  - [x] [FIX] Ruby req in `gemspec`
- [x] According to semver, how much of a version bump should the next release be?
  - [x] bumping faraday should make this at least a *minor* point release.
  - [x] officially dropping support for older Ruby versions might make this a *major* point release.

**Comment by xtagon** [comment]
What's the process for determining the minimum supported Ruby version? Do we have visibility into what Ruby versions reverse dependencies are actually using?

**Comment by josephpage** [comment]
From my research, here are the minimal ruby versions from the latest versions of the dependencies : 

- v1.9.3 Faraday - [Source (README)](https://github.com/lostisland/faraday#supported-ruby-versions)
- v2.1 JWT - [Source (gemspec)](https://github.com/jwt/ruby-jwt/blob/master/ruby-jwt.gemspec#L16)
- v1.8.7 multi_json [Source (travis.yml)](https://github.com/intridea/multi_json/blob/master/.travis.yml)
- v2.2.2 multi_xml [Source (README)](https://github.com/sferik/multi_xml/blob/master/README.md#supported-ruby-versions)
- v2.2 rack [Source (gemspec)](https://github.com/rack/rack/blob/master/rack.gemspec)

**multi_xml** should be optional, because not everybody wants to parse XML (and less APIs will be in XML in the future). **Rack** is only use for `Rack::Utils.parse_query` and `Rack::Utils.escape`. I think the `CGI` module in stdlib has alternative (`CGI::parse` and `CGI::escape`). For **JWT**, I just saw that the "Assertion" grant type is not mentioned in the final RFC.

Maybe we can hack something with the [Rubygems API](http://guides.rubygems.org/rubygems-org-api#gem-methods) to check the Gemfile of reverse dependencies.

**Comment by pboling** [comment]
- [x] `Rack::Utils.escape`
  - As of [May 2011](https://github.com/rack/rack/pull/140/files#diff-7c1a24d5b2fe58a6f925c7cacc6c55e7) Rack::Utils.escape and [continuing to today]( https://github.com/jfirebaugh/rack/blob/0700cd04c41cb2e0882521fcf2fdca877b94e0ae/lib/rack/utils.rb#L14) calls CGI.escape directly, so switching would be fine.

**Comment by josephpage** [comment]
`CGI.parse` and `Rack::Utils.parse_query` have not the same codebase but very very close. Need investigation.

Personally, I only use Ruby >= 2.4, but Rails 4.2.10 supports 1.9.3, I think we should do the best to support 1.9.3 as well.

**Comment by pboling** [comment]
Yep, let's do one more 1.X release and keep a `1.5-stable` branch for old Ruby support.

**Comment by josephpage** [comment]
Good ! 👍 
I created milestones for both releases, allowing us to tag PRs.

Ruby 2.2 will be [deprecated in March of 2018](https://www.ruby-lang.org/en/news/2017/12/14/ruby-2-2-9-released/).
I guess `2.0.0` will not be shipped before March.
We are able to be more extreme and support only 2.3+.

**Comment by galois17** [comment]
A bit late to the game but I'm going through some very old non-critical 'issues'.

**Comment by pboling** [comment]
We need to figure out where the Ruby 2.2.2 dependency is coming from in the current release, see: https://github.com/oauth-xx/oauth2/issues/264

Then we have to determine if we can remove the dependency, loosen it, or if we have already implicitly dropped support for all prior to 2.2.2.

CC @xtagon @josephpage @galois17

**Comment by galois17** [comment]
Right now, we have:
spec.add_dependency 'rack', ['>= 1.2', '< 3']

Unfortunately, Rack >= 2.0.0 requires Ruby >= 2.2.2 according to Ruby gems. So if add the proper conditionals into the gemspec file, we should be in good shape? Namely, rack >= 2.0.0 for Ruby >= 2.2.2, and rack >= 1.2 and <2.0.0 for Ruby < 2.2.2 ?

Might that all we need?

**Comment by pboling** [comment]
I wonder why bundler can't just handle that correctly.  Maybe it does now and the issue reporter was on an old version of bundler?

**Comment by grosser** [comment]
this is blocking my faraday upgrade ... can we just mark this as >= 2.2 and be done with it ?
we can later have a 1.5.x that adds back support for lower ruby versions ?

**Comment by pboling** [comment]
Master as is contains a breaking change to fix a bug that was introduced in 1.3.0, but which many people are now relying on.  So the next version will now be 2.0.0, and will retain support for older Rubies.

We plan to release 2.0.0 asap.

**Comment by pboling** [comment]
### The 2.0.0 release will continue with the current set of Rubies:

```
Ruby 1.9.3
Ruby 2.0.0
Ruby 2.1
Ruby 2.2
JRuby 1.7
JRuby 9.0
Ruby 2.3
Ruby 2.4
Ruby 2.5
JRuby 9.1
```

### With the 3.0.0 release the list of supported Rubies will shrink to:
```
Ruby 2.3
Ruby 2.4
Ruby 2.5
JRuby 9.1
```

**Comment by aried3r** [comment]
Hey @pboling! I just came across this issue when searching which gems in our project at work still require the `multi_json` gem.

It seems Ruby ships with the JSON gem since Ruby 1.9.3 and faster JSON backends like `oj` do let your override the default JSON handling, as well as shipping with a "Rails mode". 

Indeed, if you search for the use of `multi_json` in the source code, it's now mostly used in the specs.

```console
$ rg 'multi.?json' --files-with-matches --count-matches | sort
lib/oauth2/response.rb:2
oauth2.gemspec:1
spec/oauth2/access_token_spec.rb:2
spec/oauth2/client_spec.rb:8
spec/oauth2/error_spec.rb:1
spec/oauth2/response_spec.rb:6
spec/oauth2/strategy/auth_code_spec.rb:2
```

Indeed, the only usage in `response.rb` is the following:
```console
$ rg 'multi.?json' lib
lib/oauth2/response.rb
1:require 'multi_json'
125:  MultiJson.load(body) rescue body # rubocop:disable RescueModifier
```
Maybe it can be replaced by just using JSON directly? `jbuilder` recently [did the same](https://github.com/rails/jbuilder/commit/b952ae096eb1828b0fcfde06e6ba62311494ec49), see also the referenced issues for more details.

Speaking of Ruby versions, the list you posted is now over 1 year old and the only supported Ruby versions are >= 2.4 since 2.3 became EOL this year. Is it too late to speed up the process and drop old Ruby versions for the 2.x release after all?

Cheers!

**Comment by pboling** [comment]
Thanks for the research.  I will work on this soon.

**Comment by pboling** [comment]
I have reviewed multi_json, and the linked discussion in `jbuilder`.  I think `jbuilder` made a mistake, and `multi_json` appears to be the best tool here.  The entire purpose of `multi_json` is to allow you to choose which JSON engine / tool / library you want to use.

> when searching which gems in our project at work still require the multi_json gem.

I am curious to know, why were you doing this?  Is there an actual problem with the gem?  "Slow" isn't valid because when you use it you get to pick the engine you use.

Ref: https://github.com/intridea/multi_json

**Comment by aried3r** [comment]
Hey @pboling, it's been a while since I last visited this topic, but I'll try to remember our reasoning back then.

I can't speak for `jbuilder` directly, but since it's a gem meant to be used for Rails I trust that they made the right decision. We still use `oj` with it by using its [Rails integration](https://github.com/ohler55/oj/blob/develop/pages/Rails.md), so I cannot speak for pure Ruby code, but it seems [equally possible](https://github.com/ohler55/oj/blob/develop/pages/JsonGem.md).

That of course only talks about `oj`, but it seems `multi_json` is in [maintenance mode only](https://github.com/intridea/multi_json/issues/198#issuecomment-660585225) and is not being tested with the latest Ruby versions 3.0 and 3.1. That seems to include possible [security issues](https://github.com/intridea/multi_json/issues/200).
So if something should change in e.g. `oj` that breaks `multi_json` there might not be any more fixes ([like there have been in the past](https://github.com/intridea/multi_json/issues/181)).

It seems they're not against new maintainers, though:

> If you would like this library to support another Ruby version, you may volunteer to be a maintainer.

Ultimately, I don't know how many users the removal would affect nor if the one line where it's being used is in a hot path, so I'd leave that up to you since libraries have different considerations than apps.

We have since removed `multi_json` from our projects since we use `oj` directly and don't have to support other JSON backends (the other being yajl, which we considered but that also includes ["taking over" `JSON`](https://github.com/brianmario/yajl-ruby#json-gem-compatibility-api)).

**Comment by pboling** [comment]
I have reached out to @rwz to get his thoughts on this.  He is the maintainer of both `multi_json`, and `jbuilder`, the gem that removed `multi_json`... I also offerred to help with maintenance there.

The context for a Rails-excosystem gem, where dependencies can be mandated, and a Ruby gem where the ideal would be compatibility with any tools the implementer might be using, e.g. some of the JRuby-only JSON engines that `multi_json` supports.

My understanding of multi_json is that if you are using `oj`, or another JSON engine in your project, `multi_json` will detect that.  You don't need to use `multi_json` directly, or change your code.  Using it here simply allows projects to work with `oauth2` without needing to add another JSON dependency that might conflict with what they already use.

**Comment by aried3r** [comment]
That makes sense. :)

That's why I mentioned that we, in our Rails app, can freely choose while this gem might not be able to make this decision. I'm not advocating against `multi_json`, but for us it was an easy way to drop a dependency that we didn't need.

Perhaps what I'm trying to say is that without statistics and/or user feedback it's hard to say if people would notice if `multi_json` was removed and replaced by `JSON`. My _assumption_ is that this one line would not be a performance bottle-neck but again, I'm just guessing here. Or asked differently, would you add `multi_json` in a new gem in 2022.

Regarding JRuby you're correct, it seems while `gson` didn't have a release since 2013, `jrjackson`, from a quick glimpse, integrates with `multi_json` well.

To sum up, I don't feel strongly about it and I'm certainly in no position to make decision for users of this gem. Thanks for taking care of it!

> Using it here simply allows projects to work with `oauth2` without needing to add another JSON dependency that might conflict with what they already use.

I don't understand that, if a project wants to use `multi_json`with say, `oj`, they'd have to add that dependency anyway, no?

PS: Guess who's maintaining a [gem that uses `multi_json`](https://github.com/rpush/rpush/blob/995a6524e269a9a73630407d0589b4623bead9d3/lib/rpush/multi_json_helper.rb) :D

**Comment by aried3r** [comment]
> It's basiclally the same idea as `faraday`, for HTTP requests, and `multi_xml`, both of which this gem uses.

`faraday` uses `JSON` directly, but offers a `define_parser` block. See https://github.com/lostisland/faraday_middleware/issues/62#issuecomment-16471992 and `multi_xml` is an optional dependency of its (deprecated) `faraday_middleware` if you need it.

`multi_xml` also hasn't seen a new release since 2016, which is perfectly fine since some gems are just done. But they're not being tested with newer Ruby versions so personally, I'd avoid it for new projects.

I feel like I'm coming across as strongly advocating against the two `multi_*` gems, but now that I re-read my comment from 2019 I might have merely wanted to suggest to drop dependencies are part of the 2.0.0 release since it also drops old Ruby versions and it seemed like a good time to start with fewer dependencies.

Anyway, that's it from me and I want to thank you again for your work on this gem. We'll be using it, if it comes with `multi_json` or not. :)

**Comment by pboling** [comment]
I left my actual point to be inferred... should have elaborated.

> I don't understand that, if a project wants to use multi_jsonwith say, oj, they'd have to add that dependency anyway, no?

This gem *only* depends on `multi_json`.  It doesn't load or expect `json`, `json_pure`, `oj`, or anything else.

My point was...

Let's assume hypothetically that this gem begins depending directly on `json` (the stdlib C implementation) for the v2 release.

Let's assume a JRuby project _ABERNATHY_ is currently using `jrjackson`, and has been using the `oauth2` gem with `jrjackson` via `multi_json`.  This project only has one json engine loaded.

Project _ABERNATHY_ upgrades to v2, which requires `json` from stdlib, and because no C extensions on JRuby, would be using `json_pure` for OAuth2.  They now have two JSON engines loaded in their project, and they may or may not play nicely together.

This is a worse experience than having zero JSON engine dependencies in this gem, and allowing it to work with anything `multi_json` supports...

With the huge additional assumption that `multi_json` continues to be supported.

I'll look into `multi_xml` too.

**Comment by rwz** [comment]
Sorry, I haven't read much of the discussion above, but as a maintainer of MultiJson, I say that it'd make perfect sense to remove it as a dependency and replace it with native stdlib json implementation. There's no reason to keep it as a dependency unless you're targeting Ruby 1.8/1.9, which you shouldn't, probably.

**Comment by pboling** [comment]
With that insight, (thanks @rwz) I'm changing my approach on `multi_json`.  We will remove it, and have gemfiles that load up various JSON engines and test that this gem works with them.
